src/plugins/workspace/server/saved_objects/workspace_id_consumer_wrapper.ts

```diff
@@ -126,8 +131,54 @@ export class WorkspaceIdConsumerWrapper {
         }
         return wrapperOptions.client.find(finalOptions);
       },
-      bulkGet: wrapperOptions.client.bulkGet,
-      get: wrapperOptions.client.get,
+      bulkGet: async <T = unknown>(
+        objects: SavedObjectsBulkGetObject[] = [],
+        options: SavedObjectsBaseOptions = {}
+      ): Promise<SavedObjectsBulkResponse<T>> => {
+        const objectToBulkGet = await wrapperOptions.client.bulkGet<T>(objects, options);
+        const { workspaces } = this.formatWorkspaceIdParams(wrapperOptions.request, options);
+
+        // If there is a workspace in options or request, objects should belong to the workspace.
+        // Otherwise, get all the objects.
+        if (workspaces?.length === 1) {
+          const workspace = workspaces[0];
+          const savedObjects: Array<SavedObject<T>> = objectToBulkGet.saved_objects.map((object) =>
+            object.workspaces?.includes(workspace)
+              ? object
+              : {
+                  ...object,
+                  error: {
+                    ...SavedObjectsErrorHelpers.createGenericNotFoundError(object.type, object.id)
+                      .output.payload,
+                  },
+                }
+          );
+          return {
+            ...objectToBulkGet,
+            saved_objects: savedObjects,
+          };
+        }
+
+        return objectToBulkGet;
+      },
+      get: async <T = unknown>(
+        type: string,
+        id: string,
+        options: SavedObjectsBaseOptions = {}
+      ): Promise<SavedObject<T>> => {
+        const objectToGet = await wrapperOptions.client.get<T>(type, id, options);
+
+        if (!excludeTypes.includes(type)) {
+          const { workspaces } = this.formatWorkspaceIdParams(wrapperOptions.request, options);
+
+          // If there is a workspace in options or request, object should belong to the workspace.
+          if (workspaces?.length === 1 && !objectToGet.workspaces?.includes(workspaces[0])) {
+            throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+          }
```

Comment:
I am thinking we can extract the code to a method like `preflightGet(type: string, id: string)`, it can throw error if the workspace check fails and it can be used both in `get` and `bulkGet`.

Reply:
Sure, I have extract the common validate code.